simple_connection_events/simple_connection_events.cpp

```diff
@@ -0,0 +1,214 @@
+}
+
+// connection event specifics
+//////////////////////////////////////////////////////////////////////////////////////////////////////////////////
+
+
+void connected_cb(NabtoClientFuture* future, NabtoClientError ec, void* data) {
+    NabtoClientConnection* connection = (NabtoClientConnection*)data;
+    char* f;
+    if (ec != NABTO_CLIENT_EC_OK) {
+        print_connect_error(ec, connection);
+    } else if (nabto_client_connection_get_device_fingerprint_hex(connection, &f) != NABTO_CLIENT_EC_OK) {
+        std::cerr << "Could not get remote peer fingerprint" << std::endl;
+    } else {
+        std::cout << "Connected to device with fingerprint: " << std::string(f) << std::endl;
+        nabto_client_string_free(f);
+    }
+    print_connection_type(connection);
+}
+
+void die(std::string msg, cxxopts::Options options) {
+    std::cout << msg << std::endl;
+    std::cout << options.help({"", "Group"}) << std::endl;
+    exit(1);
+}
+
+static void log(const NabtoClientLogMessage* message, void* userData)
+{
+    auto now = std::chrono::system_clock::now();
+    auto ms = std::chrono::duration_cast<std::chrono::milliseconds>(now.time_since_epoch()) % 1000;
+    auto timer = std::chrono::system_clock::to_time_t(now);
+    // convert to broken time
+    std::tm bt = *std::localtime(&timer);
+
+    std::cout << std::put_time(&bt, "%H:%M:%S") << '.' << std::setfill('0') << std::setw(3) << ms.count() << " " << " " << message->module << " " << message->message << std::endl;
+}
+
+int main(int argc, char** argv) {
+    nlohmann::json opts;
+    std::string logLevel;
+
+    parse_options(argc, argv, &opts, &logLevel);
+
+    std::cout << "Nabto Client SDK Version: " << nabto_client_version() << std::endl;
+    std::cout << "opts: " << opts.dump() << std::endl;
+    std::cout << "connecting to " << opts["ProductId"].get<std::string>() << "." << opts["DeviceId"].get<std::string>() << std::endl;
+
+    NabtoClient* context = nabto_client_new();
+
+    nabto_client_set_log_callback(context, &log, NULL);
+    nabto_client_set_log_level(context, logLevel.c_str());
+
+    NabtoClientConnection* connection = nabto_client_connection_new(context);
+
+    char* privateKey;
+    nabto_client_create_private_key(context, &privateKey);
+
+    opts["PrivateKey"] = std::string(privateKey);
+    nabto_client_string_free(privateKey);
+
+    nabto_client_connection_set_options(connection, opts.dump().c_str());
+
+    init_connection_event_listener(context, connection);
+    arm_connection_event_listener(connection);
+
+    NabtoClientFuture* connectFuture = nabto_client_future_new(context);
+    nabto_client_connection_connect(connection, connectFuture);
+    nabto_client_future_set_callback(connectFuture, connected_cb, connection);
+
+    // read a character from stdin
+    std::cout << "Press enter to disconnect..." << std::endl;
+    getchar();
+
+    nabto_client_connection_close(connection, connectFuture);
```

Comment:
according to the futures docs, it is not allowed to use wait combined with a callback (callback is already set on the connect future here) - although it could be handy for synchronization at shutdown like this. So for now using a separate future for close.